test-data/unit/check-isinstance.test

```diff
@@ -1360,3 +1360,31 @@ def f(x: object) -> None:
     reveal_type(b)  # E: Revealed type is '__main__.A'
 [builtins fixtures/isinstance.pyi]
 [out]
+
+[case testIsInstanceWithUnknownType]
+from typing import Union
+def f(x: Union[int, str], typ: type) -> None:
+    if isinstance(x, (typ, int)):
+        x + 1  # E: Unsupported operand types for + (likely involving Union)
+        reveal_type(x)  # E: Revealed type is 'Union[builtins.int, builtins.str]'
+    else:
+        reveal_type(x)  # E: Revealed type is 'Union[builtins.int, builtins.str]'
```

Comment:
Here we could actually infer the type to be `builtins.str`, since we know that it can't be `builtins.int` due to the `isinstance` check.